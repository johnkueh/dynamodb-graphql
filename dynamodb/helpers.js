import * as yup from "yup";
import capitalize from "lodash/capitalize";
import moment from "moment-timezone";
import { DocumentClient } from "../lib/dynamodb-client";
import ValidationErrors from "../lib/validation-errors";

export const TableName = process.env.DYNAMODB_TABLE;
export const client = DocumentClient;

// DynamoDB Helpers
export const fetchByKey = async ({ PK, SK }) => {
  const params = {
    TableName,
    Key: {
      PK,
      SK
    }
  };
  const { Item: object } = await DocumentClient.get(params).promise();
  if (object == null) return null;
  return object;
};

export const putByKey = async ({ PK, SK, input }) => {
  const params = {
    TableName,
    Item: {
      PK,
      SK,
      id: PK,
      ...input
    }
  };
  return DocumentClient.put(params).promise();
};

export const putAndFetchByKey = async ({ PK, SK, input }) => {
  await putByKey({
    PK,
    SK,
    input
  });
  return fetchByKey({
    PK,
    SK
  });
};

export const updateByKey = async ({ PK, SK, input }) => {
  const params = {
    TableName,
    Key: {
      PK,
      SK
    },
    ...objectToExpression("UpdateExpression", input),
    ReturnValues: "ALL_NEW"
  };

  const { Attributes: object } = await DocumentClient.update(params).promise();
  return object;
};

export const deleteByKey = async ({ PK, SK }) => {
  const params = {
    TableName,
    Key: {
      PK,
      SK
    }
  };

  return DocumentClient.delete(params).promise();
};

export const objectToExpression = (type, input) => {
  const keys = Object.keys(input);
  const expressions = [];
  const ExpressionAttributeNames = {};
  const ExpressionAttributeValues = {};
  keys.forEach(key => {
    expressions.push(`#${key} = :${key}`);
    ExpressionAttributeNames[`#${key}`] = key;
    ExpressionAttributeValues[`:${key}`] = input[key];
  });

  let Expression = ``;

  switch (type) {
    case "UpdateExpression":
      Expression = `set ${expressions.join(", ")}`;
      break;
    case "KeyConditionExpression":
      Expression = `${expressions.join(" and ")}`;
      break;
    default:
  }

  return {
    [type]: Expression,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  };
};

export const validate = async (data, schema) => {
  try {
    await schema.validate(data, { abortEarly: false });
  } catch (error) {
    const { name, inner } = error;
    const errors = {};
    inner.forEach(({ path, message }) => {
      errors[path] = capitalize(message);
    });
    throw ValidationErrors(errors);
  }
};

export const validateTimezone = () =>
  yup.string().test("is-valid", "Timezone is not valid", value => {
    // Dont run the test if value is undefined
    if (value == null) return true;

    return moment.tz.names().find(name => name === value);
  });

export default {
  fetchByKey,
  putByKey,
  putAndFetchByKey,
  updateByKey,
  deleteByKey,
  objectToExpression,
  validate,
  validateTimezone
};
