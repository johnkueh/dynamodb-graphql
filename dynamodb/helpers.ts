import { DocumentClient } from "../lib/dynamodb-client";

interface ExpressionInput {
  [key: string]: string | boolean | null | undefined;
}

interface ExpressionAttributeNames {
  [key: string]: string;
}
interface ExpressionAttributeValues {
  [key: string]: string | boolean | number;
}

interface ExpressionAttributesObject {
  ExpressionAttributeNames: ExpressionAttributeNames;
  ExpressionAttributeValues: ExpressionAttributeValues;
}

interface ExpressionObject extends ExpressionAttributesObject {
  expressions: string[];
}

interface UpdateExpressionObject extends ExpressionAttributesObject {
  UpdateExpression: string;
}

interface KeyConditionExpressionObject extends ExpressionAttributesObject {
  KeyConditionExpression: string;
}

export const TableName = process.env.DYNAMODB_TABLE || "dynamodb-table";
export const client = DocumentClient;

const makeExpressions = (input: ExpressionInput): ExpressionObject => {
  const keys = Object.keys(input);
  const expressions: string[] = [];
  const ExpressionAttributeNames: ExpressionAttributeNames = {};
  const ExpressionAttributeValues: ExpressionAttributeValues = {};
  keys.forEach((key): void => {
    const inputKey = input[key];
    if (inputKey != null) {
      expressions.push(`#${key} = :${key}`);
      ExpressionAttributeNames[`#${key}`] = key;
      ExpressionAttributeValues[`:${key}`] = inputKey;
    }
  });

  return {
    expressions,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  };
};
export const makeUpdateExpression = (
  input: ExpressionInput
): UpdateExpressionObject => {
  const {
    expressions,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  } = makeExpressions(input);
  const UpdateExpression = `set ${expressions.join(", ")}`;

  return {
    UpdateExpression,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  };
};
export const makeKeyConditionExpression = (
  input: ExpressionInput
): KeyConditionExpressionObject => {
  const {
    expressions,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  } = makeExpressions(input);

  const KeyConditionExpression = `${expressions.join(" and ")}`;
  return {
    KeyConditionExpression,
    ExpressionAttributeNames,
    ExpressionAttributeValues
  };
};
