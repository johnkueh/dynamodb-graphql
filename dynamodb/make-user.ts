import jsonwebtoken from "jsonwebtoken";
import bcrypt from "bcryptjs";
import { User } from "./types";

export const makeUser = (user: User): User => {
  return {
    id: user.PK,
    teamId: user.GSI2PK,
    email: user.GSI1SK,
    jwtWithOptions: (options: Record<string, string>): string =>
      jsonwebtoken.sign(
        { id: user.PK, email: user.email, options },
        "JWTSECRET"
      ),
    jwt: jsonwebtoken.sign({ id: user.PK, email: user.GSI1SK }, "JWTSECRET"),
    validPassword: (password: string): boolean => {
      if (user.password == null) return false;
      return bcrypt.compareSync(password, user.password);
    },
    ...user
  };
};

export default {
  makeUser
};
