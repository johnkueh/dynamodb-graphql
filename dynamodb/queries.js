import uuidv4 from "uuid/v4";
import bcrypt from "bcryptjs";
import jsonwebtoken from "jsonwebtoken";
import {
  fetchByKey,
  putByKey,
  putAndFetchByKey,
  updateByKey,
  deleteByKey,
  objectToExpression,
  TableName,
  client as dynamodb
} from "./helpers";

export const Queries = {
  fetchUserById: async userId => {
    const user = await fetchByKey({
      PK: userId,
      SK: "user"
    });
    return userObject(user);
  },
  fetchUserByEmail: async email => {
    const input = {
      GSI1PK: "user",
      GSI1SK: email
    };
    const params = {
      TableName,
      IndexName: "GSI1",
      ...objectToExpression("KeyConditionExpression", input)
    };
    const { Items } = await dynamodb.query(params).promise();
    return userObject(Items[0]);
  },
  putUser: async ({ email, password, ...input }) => {
    const PK = uuidv4();
    const SK = "user";
    await putByKey({
      PK,
      SK,
      input: {
        GSI1PK: "user",
        GSI1SK: email,
        tz: "America/Los_Angeles",
        password: bcrypt.hashSync(password, 10),
        ...input
      }
    });
    return Queries.fetchUserById(PK);
  },
  updateUser: async ({ id: userId, ...input }) => {
    const user = await updateByKey({
      PK: userId,
      SK: "user",
      input
    });
    return userObject(user);
  },
  deleteUser: async ({ id: userId }) => {
    await deleteByKey(userId);
    return { id: userId };
  },
  fetchTeamById: async teamId => {
    return fetchByKey({
      PK: teamId,
      SK: "team"
    });
  },
  putTeam: async input => {
    const PK = uuidv4();
    const SK = "team";
    return putAndFetchByKey({
      PK,
      SK,
      input
    });
  },
  updateTeam: ({ id: teamId, ...input }) => {
    return updateByKey({
      PK: teamId,
      SK: "team",
      input
    });
  },
  deleteTeam: async ({ id: teamId }) => {
    await deleteByKey(teamId);
    return { id: teamId };
  },
  fetchTeamUsers: async ({ teamId }) => {
    const params = {
      TableName,
      IndexName: "GSI2",
      KeyConditionExpression: "GSI2PK = :type and GSI2SK = :teamId",
      ExpressionAttributeValues: {
        ":type": "teamUser",
        ":teamId": id
      }
    };

    const { Items } = await dynamodb.query(params).promise();
    return Items;
  },
  addUserToTeam: async ({ user, team }) => {
    const params = {
      TableName,
      Key: {
        PK: user.id,
        SK: "user"
      },
      UpdateExpression: "set #GSI2PK = :GSI2PK, #GSI2SK = :GSI2SK",
      ExpressionAttributeNames: {
        "#GSI2PK": "GSI2PK",
        "#GSI2SK": "GSI2SK"
      },
      ExpressionAttributeValues: {
        ":GSI2PK": "teamUser",
        ":GSI2SK": team.id
      },
      ReturnValues: "ALL_NEW"
    };
    const { Attributes } = await dynamodb.update(params).promise();
    user.teamId = team.id;
    return Attributes;
  }
};

// User Helpers
const userObject = user => {
  if (user == null) return null;

  return {
    jwt: jsonwebtoken.sign({ id: user.PK, email: user.email }, "JWTSECRET"),
    validPassword: password => bcrypt.compareSync(password, user.password),
    teamId: user.GSI2SK,
    email: user.GSI1SK,
    ...user
  };
};

export default {
  Queries
};
