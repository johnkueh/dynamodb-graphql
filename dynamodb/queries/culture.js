import uuidv4 from "uuid/v4";
import {
  TableName,
  makeKeyConditionExpression,
  client as DocumentClient
} from "../helpers";

export const fetchCultureById = async cultureId => {
  const params = {
    TableName,
    Key: {
      PK: cultureId,
      SK: "culture"
    }
  };
  const { Item: object } = await DocumentClient.get(params).promise();
  return object;
};

export const createCulture = async input => {
  const { name, position } = input;
  const uuid = uuidv4();
  const SK = "culture";

  const params = {
    TableName,
    Item: {
      PK: uuid,
      SK,
      id: uuid,
      GSI1PK: "culture",
      GSI1SK: `${position}|${name}|`,
      ...input
    }
  };
  await DocumentClient.put(params).promise();
  return fetchCultureById(uuid);
};
export const fetchCultures = async data => {
  const params = {
    TableName,
    IndexName: "GSI1",
    ...makeKeyConditionExpression({
      GSI1PK: "culture"
    })
  };

  const { Items } = await DocumentClient.query(params).promise();
  return Items;
};
export const fetchCultureForTeam = async teamId => {
  const params = {
    TableName,
    KeyConditionExpression: "#PK = :PK AND begins_with(#SK, :SK)",
    ExpressionAttributeNames: {
      "#PK": "PK",
      "#SK": "SK"
    },
    ExpressionAttributeValues: {
      ":PK": teamId,
      ":SK": "teamCulture|"
    }
  };

  const { Items } = await DocumentClient.query(params).promise();
  return Items;
};
export const addCultureToTeam = async input => {
  const { cultureId, teamId, position } = input;
  const culture = await fetchCultureById(cultureId);
  const PK = teamId;

  const params = {
    TableName,
    Item: {
      PK,
      SK: `teamCulture|${position}|${culture.id}`,
      id: PK,
      name: culture.name,
      ...input
    }
  };
  return DocumentClient.put(params).promise();
};
export const addCulturesToTeam = async ({ cultureIds, teamId }) => {
  await Promise.all(
    cultureIds.map((cultureId, idx) => {
      return addCultureToTeam({
        cultureId,
        teamId,
        position: idx + 1
      });
    })
  );
};
export const removeCultureFromTeam = async input => {
  const { cultureId, teamId, position } = input;
  const params = {
    TableName,
    Key: {
      PK: teamId,
      SK: `teamCulture|${position}|${cultureId}`
    }
  };

  return DocumentClient.delete(params).promise();
};
export const removeCulturesFromTeam = async teamId => {
  const teamCultureValues = await fetchCultureForTeam(teamId);
  await Promise.all(
    teamCultureValues.map(({ cultureId, position }) => {
      return removeCultureFromTeam({
        cultureId,
        position,
        teamId
      });
    })
  );
};
