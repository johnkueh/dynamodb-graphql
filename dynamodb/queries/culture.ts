import uuidv4 from "uuid/v4";
import ValidationErrors from "../../lib/validation-errors";
import {
  TableName,
  makeKeyConditionExpression,
  client as DocumentClient
} from "../helpers";
import {
  Culture,
  CultureInput,
  AddCultureInput,
  RemoveCultureInput,
  AddCulturesInput
} from "../types";

export const fetchCultureById = async (cultureId: string) => {
  const params = {
    TableName,
    Key: {
      PK: cultureId,
      SK: "culture"
    }
  };
  const { Item: object } = await DocumentClient.get(params).promise();
  return object;
};
export const createCulture = async (input: CultureInput) => {
  const { name, position } = input;
  const uuid = uuidv4();
  const SK = "culture";

  const params = {
    TableName,
    Item: {
      PK: uuid,
      SK,
      id: uuid,
      GSI1PK: "culture",
      GSI1SK: `${position}|${name}|`,
      ...input
    }
  };
  await DocumentClient.put(params).promise();
  return fetchCultureById(uuid);
};
export const fetchCultures = async () => {
  const params = {
    TableName,
    IndexName: "GSI1",
    ...makeKeyConditionExpression({
      GSI1PK: "culture"
    })
  };

  const { Items } = await DocumentClient.query(params).promise();
  return Items as Culture[];
};
export const fetchCultureForTeam = async (teamId: string) => {
  const params = {
    TableName,
    KeyConditionExpression: "#PK = :PK AND begins_with(#SK, :SK)",
    ExpressionAttributeNames: {
      "#PK": "PK",
      "#SK": "SK"
    },
    ExpressionAttributeValues: {
      ":PK": teamId,
      ":SK": "teamCulture|"
    }
  };

  const { Items } = await DocumentClient.query(params).promise();
  return Items as Culture[];
};
export const addCultureToTeam = async (input: AddCultureInput) => {
  const { cultureId, teamId, position } = input;
  const culture = await fetchCultureById(cultureId);

  if (culture == null)
    throw ValidationErrors({
      culture: "Culture not found"
    });

  const PK = teamId;

  const params = {
    TableName,
    Item: {
      PK,
      SK: `teamCulture|${position}|${culture.id}`,
      id: PK,
      name: culture.name,
      ...input
    }
  };
  return DocumentClient.put(params).promise();
};
export const addCulturesToTeam = async ({
  cultureIds,
  teamId
}: AddCulturesInput) => {
  await Promise.all(
    cultureIds.map((cultureId, idx) => {
      return addCultureToTeam({
        cultureId,
        teamId,
        position: idx + 1
      });
    })
  );
};
export const removeCultureFromTeam = async (input: RemoveCultureInput) => {
  const { cultureId, teamId, position } = input;
  const params = {
    TableName,
    Key: {
      PK: teamId,
      SK: `teamCulture|${position}|${cultureId}`
    }
  };

  return DocumentClient.delete(params).promise();
};
export const removeCulturesFromTeam = async (teamId: string) => {
  const teamCultureValues = await fetchCultureForTeam(teamId);

  if (teamCultureValues == null) return;

  await Promise.all(
    teamCultureValues.map(({ cultureId, position }) => {
      return removeCultureFromTeam({
        cultureId,
        position,
        teamId
      });
    })
  );
};
