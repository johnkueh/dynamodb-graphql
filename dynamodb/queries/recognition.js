import uuidv4 from "uuid/v4";
import moment from "moment";
import {
  fetchByKey,
  putByKey,
  putAndFetchByKey,
  updateByKey,
  deleteByKey,
  TableName,
  client as DocumentClient
} from "../helpers";

export const Queries = {
  putCulture: async input => {
    const { name, position } = input;
    const PK = uuidv4();
    const SK = "culture";
    return putAndFetchByKey({
      PK,
      SK,
      input: {
        GSI1PK: "culture",
        GSI1SK: `${position}|${name}|`,
        ...input
      }
    });
  },
  fetchCultureById: async cultureId => {
    return fetchByKey({
      PK: cultureId,
      SK: "culture"
    });
  },
  fetchCultures: async data => {
    const params = {
      TableName,
      IndexName: "GSI1",
      KeyConditionExpression: "GSI1PK = :type",
      ExpressionAttributeValues: {
        ":type": "culture"
      }
    };

    const { Items } = await DocumentClient.query(params).promise();
    return Items;
  },
  culture: {
    fetchForTeam: async teamId => {
      const params = {
        TableName,
        KeyConditionExpression: "#PK = :PK AND begins_with(#SK, :SK)",
        ExpressionAttributeNames: {
          "#PK": "PK",
          "#SK": "SK"
        },
        ExpressionAttributeValues: {
          ":PK": teamId,
          ":SK": "teamCulture|"
        }
      };

      const { Items } = await DocumentClient.query(params).promise();
      return Items;
    },
    addToTeam: async input => {
      const { cultureId, teamId, position } = input;
      const culture = await Queries.fetchCultureById(cultureId);
      return putByKey({
        PK: teamId,
        SK: `teamCulture|${position}|${culture.id}`,
        input: {
          cultureId: culture.id,
          name: culture.name,
          position
        }
      });
    },
    addAllToTeam: async ({ cultureIds, teamId }) => {
      await Promise.all(
        cultureIds.map((cultureId, idx) => {
          return Queries.culture.addToTeam({
            cultureId,
            teamId,
            position: idx + 1
          });
        })
      );
    },
    removeAllFromTeam: async teamId => {
      const teamCultureValues = await Queries.culture.fetchForTeam(teamId);
      await Promise.all(
        teamCultureValues.map(({ cultureId, position }) => {
          return Queries.culture.removeFromTeam({
            cultureId,
            position,
            teamId
          });
        })
      );
    },
    removeFromTeam: async input => {
      const { cultureId, teamId, position } = input;
      return deleteByKey({
        PK: teamId,
        SK: `teamCulture|${position}|${cultureId}`
      });
    }
  },
  responses: {
    // fetchCountsForAllCultures: async () => {},
    // fetchCountsForTeamCultures: async teamId => {},
    // fetchCountsForUserCultures: async userId => {},
    // fetchCountForCultureId: async cultureId => {},

    // fetchCountsForAllEmojis: async () => {},
    // fetchCountsForTeamEmojis: async teamId => {},
    // fetchCountsForUserEmojis: async userId => {},
    // fetchCountForEmoji: async emoji => {}
    fetchById: async () => {},
    put: async data => {
      const { userId, teamId, sentAt, ...input } = data;
      const PK = uuidv4();
      const SK = `response`;
      return putAndFetchByKey({
        PK,
        SK,
        input: {
          GSI1PK: `response_${teamId}`,
          GSI1SK: sentAt,
          GSI2PK: `response_${userId}`,
          GSI2SK: sentAt,
          sentAt,
          userId,
          teamId,
          ...input
        }
      });
    },
    update: async () => {},
    user: {
      fetchByDateRange: async ({ userId, fromDate, toDate }) => {}
    },
    team: {
      fetchByDateRange: async ({ teamId, fromDate, toDate }) => {
        const params = {
          TableName,
          IndexName: "GSI1",
          KeyConditionExpression:
            "GSI1PK = :teamResponseId AND GSI1SK between :fromDate AND :toDate",
          ExpressionAttributeValues: {
            ":teamResponseId": `response_${teamId}`,
            ":fromDate": fromDate,
            ":toDate": toDate
          }
        };

        const { Items } = await DocumentClient.query(params).promise();
        // console.log(Items);
        return Items;
      }
    }
  }
};

export default {
  Queries
};
