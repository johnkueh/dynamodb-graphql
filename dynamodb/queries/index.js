import { UserQueries } from "./user";
import { TeamQueries } from "./team";
import { Queries as RecognitionQueries } from "./recognition";

export const Queries = {
  ...UserQueries,
  ...TeamQueries,
  ...RecognitionQueries
};

export default {
  Queries
};
