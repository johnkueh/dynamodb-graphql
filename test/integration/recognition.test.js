import moment from "moment";
import jsonwebtoken from "jsonwebtoken";
import "../support/globals";
import { query as performQuery } from "../support/apollo-test-helper";
import { Queries } from "../../dynamodb/queries";

describe("Fetching culture values", () => {
  const query = `
    query {
      cultureValues {
        name
      }
    }
  `;
  beforeEach(async () => {
    await Queries.createCulture({ name: "Teamwork", position: 1 });
    await Queries.createCulture({ name: "Integrity", position: 2 });
    await Queries.createCulture({ name: "Helpful", position: 3 });
  });

  it("can fetch culture values without auth", async () => {
    const res = await performQuery({
      query
    });

    expect(res).toMatchSnapshot();
  });
});

describe("viewing responses", () => {
  let user;
  let user2;
  const query = `
    query($input: ResponsesFilterInput!) {
      responses(input: $input) {
        feeling
        user {
          email
        }
      }
    }
    `;

  beforeEach(async () => {
    user = await Queries.createUserWithTeam({
      name: "Darth Vader",
      email: "darth@vader.com",
      password: "password",
      teamName: "Star Wars"
    });

    await Queries.createResponse({
      sentAt: moment("2018-05-10").toISOString(),
      submittedAt: moment("2018-05-10").toISOString(),
      userId: user.id,
      teamId: user.teamId,
      feeling: "HAPPY"
    });

    user2 = await Queries.createUser({
      name: "Anakin Skywalker",
      email: "anakin@skywalker.com",
      password: "password"
    });

    await Queries.addUserToTeam({
      user: user2,
      team: user.team
    });

    await Queries.createResponse({
      sentAt: moment("2018-05-11").toISOString(),
      submittedAt: moment("2018-05-11").toISOString(),
      userId: user2.id,
      teamId: user2.team.id,
      feeling: "HAPPY"
    });
  });

  it("not able to list responses if not logged in", async () => {
    const res = await performQuery({
      query,
      variables: {
        input: {
          from: "",
          to: ""
        }
      }
    });

    expect(JSON.stringify(res)).toMatchSnapshot();
  });

  it("is able to list responses belonging to team", async () => {
    // This user's response shouldnt appear because its not in the user's team
    const otherTeamUser = await Queries.createUserWithTeam({
      name: "Tom n Jerry",
      email: "tom@jerry.com",
      password: "password",
      teamName: "Disney"
    });

    await Queries.createResponse({
      sentAt: moment("2018-05-10").toISOString(),
      submittedAt: moment("2018-05-10").toISOString(),
      userId: otherTeamUser.id,
      teamId: otherTeamUser.team.id,
      feeling: "HAPPY"
    });

    const res = await performQuery({
      context: { user },
      query,
      variables: {
        input: {
          from: moment("2018-05-01").toISOString(),
          to: moment("2018-06-01").toISOString()
        }
      }
    });

    expect(res).toMatchSnapshot();
  });

  it("only show responses in the filtered dates", async () => {
    // This response entry shouldnt appear because its not in the filtered month
    await Queries.createResponse({
      sentAt: moment("2018-01-10").toISOString(),
      submittedAt: moment("2018-01-10").toISOString(),
      userId: user.id,
      teamId: user.team.id,
      feeling: "SAD"
    });

    const res = await performQuery({
      context: { user },
      query,
      variables: {
        input: {
          from: moment("2018-05-01").toISOString(),
          to: moment("2018-06-01").toISOString()
        }
      }
    });

    expect(res).toMatchSnapshot();
  });
});

describe("submitting responses", () => {
  let user;
  let response;
  const query = `
    mutation($input: UpdateResponseInput!) {
      updateResponse(input: $input) {
        response {
          feeling
          submittedAt
        }
        jwt
      }
    }
  `;
  beforeEach(async () => {
    user = await Queries.createUserWithTeam({
      name: "Darth Vader",
      email: "darth@vader.com",
      password: "password",
      teamName: "Star Wars"
    });
    response = await Queries.createResponse({
      sentAt: moment("2018-05-10").toISOString(),
      userId: user.id,
      teamId: user.teamId,
      feeling: "HAPPY"
    });
  });

  it("is able to update response without logged in", async () => {
    expect(response.submittedAt).toBeUndefined();

    const res = await performQuery({
      query,
      variables: {
        input: {
          id: response.id,
          feeling: "HAPPY"
        }
      }
    });

    const decoded = jsonwebtoken.verify(
      res.data.updateResponse.jwt,
      "JWTSECRET"
    );

    expect(decoded).toEqual({
      id: expect.any(String),
      iat: expect.any(Number),
      email: user.email,
      options: {
        expiresIn: "1h"
      }
    });

    expect(res).toMatchSnapshot({
      data: {
        updateResponse: {
          jwt: expect.any(String),
          response: {
            submittedAt: expect.any(String)
          }
        }
      }
    });
  });

  it("is not able to update response with invalid id", async () => {
    const res = await performQuery({
      query,
      variables: {
        input: {
          id: "12312313",
          feeling: "HAPPY"
        }
      }
    });

    expect(JSON.stringify(res)).toMatchSnapshot();
  });

  it("is not able to update response without feeling", async () => {
    const res = await performQuery({
      query,
      variables: {
        input: {
          id: response.id,
          feeling: ""
        }
      }
    });

    expect(JSON.stringify(res)).toMatchSnapshot();
  });

  it("is not able to update response with invalid feeling", async () => {
    const res = await performQuery({
      query,
      variables: {
        input: {
          id: response.id,
          feeling: "FU"
        }
      }
    });

    expect(JSON.stringify(res)).toMatchSnapshot();
  });
});
