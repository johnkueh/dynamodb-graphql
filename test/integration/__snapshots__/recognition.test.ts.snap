// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetching culture values can fetch culture values without auth 1`] = `
Object {
  "data": Object {
    "cultureValues": Array [
      Object {
        "name": "Teamwork",
      },
      Object {
        "name": "Integrity",
      },
      Object {
        "name": "Helpful",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`submitting responses is able to update response without logged in 1`] = `
Object {
  "data": Object {
    "updateResponse": Object {
      "jwt": Any<String>,
      "response": Object {
        "feeling": "HAPPY",
        "submittedAt": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`submitting responses is not able to update response with invalid feeling 1`] = `"{\\"http\\":{\\"headers\\":{}},\\"errors\\":[{\\"message\\":\\"ValidationError\\",\\"locations\\":[{\\"line\\":3,\\"column\\":7}],\\"path\\":[\\"updateResponse\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\",\\"exception\\":{\\"errors\\":{\\"feeling\\":\\"Feeling must be one of the following values: happy, neutral, sad\\"}}}}],\\"data\\":null}"`;

exports[`submitting responses is not able to update response with invalid id 1`] = `"{\\"http\\":{\\"headers\\":{}},\\"errors\\":[{\\"message\\":\\"ValidationError\\",\\"locations\\":[{\\"line\\":3,\\"column\\":7}],\\"path\\":[\\"updateResponse\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\",\\"exception\\":{\\"errors\\":{\\"auth\\":\\"You are not authorized to perform this action\\"}}}}],\\"data\\":null}"`;

exports[`submitting responses is not able to update response without feeling 1`] = `"{\\"http\\":{\\"headers\\":{}},\\"errors\\":[{\\"message\\":\\"ValidationError\\",\\"locations\\":[{\\"line\\":3,\\"column\\":7}],\\"path\\":[\\"updateResponse\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\",\\"exception\\":{\\"errors\\":{\\"feeling\\":\\"Feeling must be one of the following values: happy, neutral, sad\\"}}}}],\\"data\\":null}"`;

exports[`viewing responses is able to list responses belonging to team 1`] = `
Object {
  "data": Object {
    "responses": Array [
      Object {
        "feeling": "HAPPY",
        "user": Object {
          "email": "darth@vader.com",
        },
      },
      Object {
        "feeling": "HAPPY",
        "user": Object {
          "email": "anakin@skywalker.com",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`viewing responses not able to list responses if not logged in 1`] = `"{\\"http\\":{\\"headers\\":{}},\\"errors\\":[{\\"message\\":\\"ValidationError\\",\\"locations\\":[{\\"line\\":3,\\"column\\":7}],\\"path\\":[\\"responses\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\",\\"exception\\":{\\"errors\\":{\\"auth\\":\\"You are not authorized to perform this action\\"}}}}],\\"data\\":null}"`;

exports[`viewing responses only show responses in the filtered dates 1`] = `
Object {
  "data": Object {
    "responses": Array [
      Object {
        "feeling": "HAPPY",
        "user": Object {
          "email": "darth@vader.com",
        },
      },
      Object {
        "feeling": "HAPPY",
        "user": Object {
          "email": "anakin@skywalker.com",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
